File: native-ios/RideTrackerNative/HUDConfiguration.swift
import Foundation

struct HUDElementConfiguration: Codable, Equatable {
    var visible: Bool = true
    var x: Double
    var y: Double
    var width: Double
    var height: Double
    var scale: Double = 1
    var opacity: Double = 1
    var fontScale: Double = 1
}

struct HUDProfileConfiguration: Codable, Equatable {
    var panelOpacity: Double = 0.86
    var globalOpacity: Double = 1
    var fontScale: Double = 1
    var fontFamily: String = "system"
    var elements: [String: HUDElementConfiguration]
}

struct HUDWatermarkConfiguration: Codable, Equatable {
    var enabled: Bool = false
    var imageFilename: String?
    var x: Double = 0.82
    var y: Double = 0.04
    var width: Double = 0.14
    var opacity: Double = 0.65
}

struct HUDConfiguration: Codable, Equatable {
    var version: String = "1.0.0"
    var profiles: [String: HUDProfileConfiguration]
    var watermark = HUDWatermarkConfiguration()

    static func defaults(from overlay: OverlayConfiguration) -> HUDConfiguration {
        func profile(_ layout: [String: [Double]]) -> HUDProfileConfiguration {
            let elements = layout.reduce(into: [String: HUDElementConfiguration]()) { result, pair in
                guard pair.value.count == 4 else { return }
                result[pair.key] = HUDElementConfiguration(
                    x: pair.value[0], y: pair.value[1], width: pair.value[2], height: pair.value[3]
                )
            }
            return HUDProfileConfiguration(elements: elements)
        }
        return HUDConfiguration(profiles: [
            "landscape": profile(overlay.layouts["landscape"] ?? [:]),
            "portrait": profile(overlay.layouts["portrait"] ?? [:])
        ])
    }
}

File: native-android/app/src/main/java/de/ridetracker/overlay/HudConfiguration.kt
package de.ridetracker.overlay

data class HudElementConfiguration(
    val visible: Boolean = true,
    val x: Double,
    val y: Double,
    val width: Double,
    val height: Double,
    val scale: Double = 1.0,
    val opacity: Double = 1.0,
    val fontScale: Double = 1.0,
)

data class HudProfileConfiguration(
    val panelOpacity: Double = 0.86,
    val globalOpacity: Double = 1.0,
    val fontScale: Double = 1.0,
    val fontFamily: String = "system",
    val elements: Map<String, HudElementConfiguration>,
)

data class HudWatermarkConfiguration(
    val enabled: Boolean = false,
    val imageFilename: String? = null,
    val x: Double = 0.82,
    val y: Double = 0.04,
    val width: Double = 0.14,
    val opacity: Double = 0.65,
)

data class HudConfiguration(
    val version: String = "1.0.0",
    val profiles: Map<String, HudProfileConfiguration>,
    val watermark: HudWatermarkConfiguration = HudWatermarkConfiguration(),
) {
    companion object {
        fun defaults(overlay: OverlayConfiguration): HudConfiguration {
            fun profile(layout: Map<String, List<Double>>) = HudProfileConfiguration(
                elements = layout.mapNotNull { (key, value) ->
                    if (value.size != 4) null else key to HudElementConfiguration(
                        x = value[0], y = value[1], width = value[2], height = value[3],
                    )
                }.toMap(),
            )
            return HudConfiguration(
                profiles = mapOf(
                    "landscape" to profile(overlay.layouts["landscape"].orEmpty()),
                    "portrait" to profile(overlay.layouts["portrait"].orEmpty()),
                )
            )
        }
    }
}
